trace/trace.go

```diff
@@ -93,29 +146,41 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
 		writeFields(&buffer, t.fields)
 		buffer.WriteString(" ")
 	}
-	buffer.WriteString(fmt.Sprintf("(started: %v) (total time: %v):\n", t.startTime, totalTime))
-	lastStepTime := t.startTime
-	for _, step := range t.steps {
-		stepDuration := step.stepTime.Sub(lastStepTime)
-		if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4) {
-			buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] ", tracenum, step.stepTime.Sub(t.startTime), stepDuration))
-			buffer.WriteString(step.msg)
-			if len(step.fields) > 0 {
-				buffer.WriteString(" ")
-				writeFields(&buffer, step.fields)
-			}
-			buffer.WriteString("\n")
-		}
-		lastStepTime = step.stepTime
-	}
+	buffer.WriteString(fmt.Sprintf("(%v) (total time: %vms):", t.startTime.Format("02-Jan-2006 15:04:00.000"), totalTime.Milliseconds()))
+	lastStepTime := writeTrace(&buffer, t, fmt.Sprintf("\nTrace[%d]: ", tracenum), stepThreshold)
 	stepDuration := endTime.Sub(lastStepTime)
 	if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4) {
-		buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] END\n", tracenum, endTime.Sub(t.startTime), stepDuration))
+		buffer.WriteString(fmt.Sprintf("\nTrace[%d]: [%v] [%v] END\n", tracenum, endTime.Sub(t.startTime), stepDuration))
 	}
 
 	klog.Info(buffer.String())
 }
 
+func writeTrace(b *bytes.Buffer, t *Trace, formatter string, stepThreshold time.Duration) time.Time {
+	lastStepTime := t.startTime
+	stepAndTraces := t.sortStepTrace()
+	if len(stepAndTraces) == 0 {
+		return lastStepTime
+	}
+	for _, stepOrTrace := range stepAndTraces {
+		switch s := stepOrTrace.(type) {
```

Comment:
An alternative would be to add a `writeStep` function to the `stepTraceTime` interface (and maybe rename the interface to something more general), and then instead of having a switch here, just call `writeStep` which will call the appropriate code on both `Trace` and `traceStep`.